### PR TITLE
2544 terraform s3 policy

### PR DESCRIPTION
**Pull request by m-s-austin**
This code adds a new provider to terraform that allows you to add a Policy to an s3 bucket like this:

resource "aws_s3_bucket_policy" "test_policy" {
name = "Bucket Test Policy"
bucket = "${aws_s3_bucket.b.id}"
policy = <<EOF
{
"Id": "Policy1431447664360",
"Version": "2012-10-17",
"Statement": [
{
"Sid": "Stmt1431447659873",
"Action": [
"s3:GetObject"
],
"Effect": "Allow",
"Resource": "arn:aws:s3:::my_tf_test_bucket/",
"Principal": ""
}
]
}
EOF
}

### CODE REVIEW

**Comment by chimmelb** [comment]
+1
Admittedly I don't know Go. The code looks clean enough.

**Comment by tehnorm** [comment]
Did we figure out the unit testing process? Would love to see working unit tests with this before we merge it down.

**Comment by m-s-austin** [comment]
@tehnorm They have "acceptance tests" but not "unit tests" the acceptance tests still run, I can work on figuring out how to write more. They appear to be more of an integration type test than a unit test, in that they actually create / destroy live resources on the services. 

It's going to take at least a day or two of research to understand how to write my own. However, I might be able to get some feedback from hashicorp. that might help.

File: builtin/providers/aws/resource_aws_s3_bucket_policy.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		},
+	}
+}
+
+func resourceAwsS3BucketPolicyPut(d *schema.ResourceData, meta interface{}) error {
+
+	s3conn := meta.(*AWSClient).s3conn
+	bucket := d.Get("bucket").(string)
+	policy := d.Get("policy").(string)
+	name := d.Get("name").(string)
```
Comment (Carrington) [inline builtin/providers/aws/resource_aws_s3_bucket_policy.go:43]: So I also do not know Go, but would it be possible to give these variables some more descriptive names? What is 'd', for instance?

Comment (m-s-austin) [inline builtin/providers/aws/resource_aws_s3_bucket_policy.go:43]: It's basically their standard:
https://godoc.org/github.com/hashicorp/terraform/helper/schema#pkg-index

d *ResourceData technically if I was going strictly by their practices I probably should have stuck with m for meta. 

I agree it's not very descriptive.

**Comment by m-s-austin** [comment]
I will probably be closing this in favor of 
https://github.com/hashicorp/terraform/pull/1992

I have been testing it and it seems to have all the same functionality as well as already having tests.

**Comment by m-s-austin** [comment]
1992 has already been merged into master on hashicorp.

**Pull request closed without merging**